test/reshape.jl

```diff
@@ -386,4 +396,64 @@ end
     @test df_flat_cat.b isa CategoricalArray
 end
 
+@testset "test RepeatedVector for categorical" begin
+    v = categorical(["a", "b", "c"], ordered=true)
+    levels!(v, ["b", "c", "a"])
+    rv = DataFrames.RepeatedVector(v, 1, 1)
+    @test isordered(v)
+    # uncomment after CategoricalArrays.jl is fixed
+    # @test isordered(categorical(v))
+    @test levels(v) == ["b", "c", "a"]
+    @test levels(categorical(v)) == ["b", "c", "a"]
+
+    v = categorical(["a", "b", "c"])
+    levels!(v, ["b", "c", "a"])
+    rv = DataFrames.RepeatedVector(v, 1, 1)
+    @test !isordered(v)
+    # uncomment after CategoricalArrays.jl is fixed
+    # @test !isordered(categorical(v))
+    @test levels(v) == ["b", "c", "a"]
+    @test levels(categorical(v)) == ["b", "c", "a"]
+end
+
+@testset "stack categorical test" begin
+    d1 = DataFrame(a = repeat([1:3;], inner = [4]),
+                   b = repeat([1:4;], inner = [3]),
+                   c = randn(12),
+                   d = randn(12),
+                   e = map(string, 'a':'l'))
+    d1s = stack(d1, [:d, :c])
+    @test d1s.variable isa CategoricalVector{String}
+    @test levels(d1s.variable) == ["d", "c"]
+    d1s = stack(d1, [:d, :c], view=true)
+    @test d1s.variable isa DataFrames.RepeatedVector{<:CategoricalString}
+    @test levels(d1s.variable) == ["d", "c"]
+    @test d1s[:, 1] isa CategoricalVector{String}
+    @test levels(d1s[:, 1]) == ["d", "c"]
+
+    d1s = stack(d1, [:d, :c], variable_eltype=String)
+    @test d1s.variable isa PooledVector{String}
+    @test levels(d1s.variable) == ["c", "d"]
+    d1s = stack(d1, [:d, :c], view=true, variable_eltype=String)
+    @test d1s.variable isa DataFrames.RepeatedVector{String}
+    @test levels(d1s.variable) == ["c", "d"]
+    @test d1s[:, 1] isa Vector{String}
+    @test levels(d1s[:, 1]) == ["c", "d"]
+
+    d1s = stack(d1, [:d, :c], variable_eltype=Symbol)
+    @test d1s.variable isa Vector{Symbol}
+    @test levels(d1s.variable) == [:c, :d]
+    d1s = stack(d1, [:d, :c], view=true, variable_eltype=Symbol)
+    @test d1s.variable isa DataFrames.RepeatedVector{Symbol}
+    @test levels(d1s.variable) == [:c, :d]
+    @test d1s[:, 1] isa Vector{Symbol}
+    @test levels(d1s[:, 1]) == [:c, :d]
+
+    d2 = categorical(d1, :)
```

Comment:
I have applied all your comments - thank you. In particular here I have added the test we were talking about (I hope I understood you correctly what you wanted us to test here?)

Reply:
Yes, that's it. Though it would be even better if you tested that when columns have the same levels their order is preserved (as well as `isordered`).

Reply:
added. Note though that `:value` column is in this case merging the original levels (as it combines two vectors in our case), but I think it is OK (normally `:value` will not be categorical, but I just wanted to make sure we are OK in this case also).

Reply:
Hmm, but if you keep levels of `:c` and `:d` in their default order and without marking them as ordered, you can't check that order is preserved, right? My idea was that you can make both variables use the same levels (including unused ones).

Reply:
I am not fully sure what you mean by "you can't check that order is preserved" and by "both variables use the same levels (including unused ones)" (in the output we have a single column.).

what `stack` does essentially is that it calls `vcat(d2.c, d2.d)` and all that happens is exactly what `vcat` provides. I think that if you feel that we do not get what we like here then `vcat` implementation for categorical arrays should be changed - not the code of `stack` (but maybe I misunderstand something).

In particular note that if you `vcat` two ordered categorical arrays they become unordered and their levels are just added sequentially:
```
julia> x = categorical(1:3, ordered=true)
3-element CategoricalArray{Int64,1,UInt32}:
 1
 2
 3

julia> y = categorical(4:6, ordered=true)
3-element CategoricalArray{Int64,1,UInt32}:
 4
 5
 6

julia> isordered(vcat(x, y))
false

julia> levels(vcat(x, y))
6-element Array{Int64,1}:
 1
 2
 3
 4
 5
 6

julia> levels(vcat(y, x))
6-element Array{Int64,1}:
 4
 5
 6
 1
 2
 3
```
What other behavior would you expect here?

Reply:
I was thinking about something like this:
```julia
julia> x = levels!(categorical(1:3, ordered=true), collect(6:-1:1))
3-element CategoricalArray{Int64,1,UInt32}:
 1
 2
 3

julia> y = levels!(categorical(4:6, ordered=true), collect(6:-1:1))
3-element CategoricalArray{Int64,1,UInt32}:
 4
 5
 6

julia> isordered(vcat(x, y))
true

julia> levels(vcat(x, y))
6-element Array{Int64,1}:
 6
 5
 4
 3
 2
 1
```

That's just the behavior of `vcat`, but it may still be worth checking since in other places we use other tricks to allocate an array and fill it manually, which may break levels.

Reply:
sure - done! Thank you.